Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,456 @@
+
+/-!
+# Coxeter Systems
+
+This file defines Coxeter systems and Coxeter groups.
+
+A Coxeter system is a pair `(W, S)` where `W` is a group generated by a set of
+reflections (involutions) `S = {s₁,s₂,...,sₙ}`, subject to relations determined
+by a Coxeter matrix `M = (mᵢⱼ)`.  The Coxeter matrix is a symmetric matrix with entries
+`mᵢⱼ` representing the order of the product `sᵢsⱼ` for `i ≠ j` and `mᵢᵢ = 1`.
+
+When `(W, S)` is a Coxeter system, one also says, by abuse of language, that `W` is a
+Coxeter group.  A Coxeter group `W` is determined by the presentation
+`W = ⟨s₁,s₂,...,sₙ | ∀ i j, (sᵢsⱼ)^mᵢⱼ = 1⟩`, where `1` is the identity element of `W`.
+
+The finite Coxeter groups are classified (TODO) as the four infinite families:
+
+* `Aₙ, Bₙ, Dₙ, I₂ₘ`
+
+And the exceptional systems:
+
+* `E₆, E₇, E₈, F₄, G₂, H₃, H₄`
+
+## Implementation details
+
+In this file a Coxeter system, designated as `CoxeterSystem W`, is implemented as a
+structure which effectively records the isomorphism between a group `W` and the corresponding
+group presentation derived from a Coxeter matrix.  From another perspective, it serves as a
+set of generators for `W`, tailored to the underlying type of `M`, while ensuring compliance
+with the relations specified by the Coxeter matrix `M`.
+
+A type class `IsCoxeterGroup` is introduced, for groups that are isomorphic to a group
+presentation corresponding to a Coxeter matrix which is registered in a Coxeter system.
+
+## Main definitions
+
+* `Matrix.IsCoxeter` : A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries
+  equal to one and off-diagonal entries distinct from one.
+* `Matrix.CoxeterGroup` : The group presentation corresponding to a Coxeter matrix.
+* `CoxeterSystem` : A structure recording the isomorphism between a group `W` and the group
+  presentation corresponding to a Coxeter matrix, i.e. `Matrix.CoxeterGroup M`.
+* `equivCoxeterGroup` : Coxeter groups of isomorphic types are isomorphic.
+* `IsCoxeterGroup` : A group is a Coxeter group if it is registered in a Coxeter system.
+* `CoxeterMatrix.Aₙ`
+* `CoxeterMatrix.Bₙ`
+* `CoxeterMatrix.Dₙ`
+* `CoxeterMatrix.I₂ₘ`
+* `CoxeterMatrix.E₆`
+* `CoxeterMatrix.E₇`
+* `CoxeterMatrix.E₈`
+* `CoxeterMatrix.F₄`
+* `CoxeterMatrix.G₂`
+* `CoxeterMatrix.H₃`
+* `CoxeterMatrix.H₄`
+
+## References
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 4--6*](bourbaki1968) chapter IV
+  pages 4--5, 13--15
+
+## TODO
+
+* The canonical map from the type to the Coxeter group `W` is an injection.
+
+* A group `W` registered in a Coxeter system is a Coxeter group.
+
+* A Coxeter group is an instance of `IsCoxeterGroup`.
+
+## Tags
+
+coxeter system, coxeter group
+-/
+
+
+universe u
+
+noncomputable section
+
+variable {B : Type*} [DecidableEq B]
+
+variable (M : Matrix B B ℕ)
+
+/-- A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries equal to one
+and off-diagonal entries distinct from one. -/
+class Matrix.IsCoxeter : Prop where
```

Comment:
Is there a reason this is a `class` and not a `structure`? If so, can you add it to the docstring?

Reply:
Not that I know and I actually don't see too many future places where this would be needed (Lean's type inference system) so will change it to a structure.